File: Go/0017-letter-combinations-of-a-phone-number.go
func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return []string{}
	}

	// Map digits to their corresponding letters
	phoneMap := map[byte]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}

	var letterMappings []string
	var result []string

	// Store the letter mappings for each digit
	for i := 0; i < len(digits); i++ {
		letterMappings = append(letterMappings, phoneMap[digits[i]])
	}

	// Helper function for backtracking
	var generateCombinations func(index int, currentString string)
	generateCombinations = func(index int, currentString string) {
		// Base case: if we've processed all digits
		if index == len(letterMappings) {
			result = append(result, currentString)
			return
		}

		// Get all letters for the current digit
		letters := letterMappings[index]

		// Try each letter
		for i := 0; i < len(letters); i++ {
			// Add current letter to the combination and move to next digit
			generateCombinations(index+1, currentString+string(letters[i]))
		}
	}

	// Start recursive backtracking
	generateCombinations(0, "")

	return result
}

File: Java/0017-letter-combinations-of-a-phone-number.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    private Map<Character, String> phoneMap = new HashMap<>() {{
        put('2', "abc");
        put('3', "def");
        put('4', "ghi");
        put('5', "jkl");
        put('6', "mno");
        put('7', "pqrs");
        put('8', "tuv");
        put('9', "wxyz");
    }};
    
    private List<String> letterMappings = new ArrayList<>();
    private List<String> result = new ArrayList<>();
    
    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty()) {
            return result;
        }
        
        // Clear lists in case the method is called multiple times
        letterMappings.clear();
        result.clear();
        
        // Store the letter mappings for each digit
        for (char digit : digits.toCharArray()) {
            letterMappings.add(phoneMap.get(digit));
        }
        
        // Start recursive backtracking
        generateCombinations(0, "");
        
        return result;
    }
    
    private void generateCombinations(int index, String currentString) {
        // Base case: if we've processed all digits
        if (index == letterMappings.size()) {
            result.add(currentString);
            return;
        }
        
        // Get all letters for the current digit
        String letters = letterMappings.get(index);
        
        // Try each letter
        for (char letter : letters.toCharArray()) {
            // Add current letter to the combination and move to next digit
            generateCombinations(index + 1, currentString + letter);
        }
    }
}